python/ray/serve/api.py

```diff
@@ -427,65 +429,155 @@ def decorator(_func_or_class):
+            raise TypeError(
+                "`serve.run` expects an `Application` returned by `Deployment.bind()`."
+            )
+
+        validate_route_prefix(t.route_prefix)
+
+        built_apps.append(
+            build_app(
+                t.target,
+                name=t.name,
+                route_prefix=t.route_prefix,
+                logging_config=t.logging_config,
+                make_deployment_handle=make_local_deployment_handle
+                if _local_testing_mode
+                else None,
+            )
+        )
 
     if _local_testing_mode:
         configure_component_logger(
             component_name="local_test",
             component_id="-",
-            logging_config=logging_config or LoggingConfig(),
+            logging_config=t.logging_config  # implicitly uses the last target
+            or LoggingConfig(),
             stream_handler_only=True,
         )
-        built_app = build_app(
-            target,
-            name=name,
-            make_deployment_handle=make_local_deployment_handle,
-        )
-        handle = built_app.deployment_handles[built_app.ingress_deployment_name]
+        return [b.deployment_handles[b.ingress_deployment_name] for b in built_apps]
     else:
         client = _private_api.serve_start(
             http_options={"location": "EveryNode"},
-            global_logging_config=logging_config,
+            global_logging_config=t.logging_config,  # implicitly uses the last target
         )
+
         # Record after Ray has been started.
         ServeUsageTag.API_VERSION.record("v2")
-        handle = client.deploy_application(
-            build_app(target, name=name),
-            blocking=_blocking,
-            route_prefix=route_prefix,
-            logging_config=logging_config,
+
+        return client.deploy_applications(
+            built_apps,
+            wait_for_ingress_deployment_creation=wait_for_ingress_deployment_creation,
+            wait_for_applications_running=wait_for_applications_running,
         )
 
-    return handle
+
+@PublicAPI(stability="stable")
+def _run(
+    target: Application,
+    *,
+    _blocking: bool = True,
+    name: str = SERVE_DEFAULT_APP_NAME,
+    route_prefix: Optional[str] = "/",
+    logging_config: Optional[Union[Dict, LoggingConfig]] = None,
+    _local_testing_mode: bool = False,
+) -> DeploymentHandle:
+    """Run an application and return a handle to its ingress deployment.
+
+    This is only used internally with the _blocking not totally blocking the following
+    code indefinitely until Ctrl-C'd.
+    """
+    return _run_many(
+        [
+            RunTarget(
+                target=target,
+                name=name,
+                route_prefix=route_prefix,
+                logging_config=logging_config,
+            )
+        ],
+        wait_for_applications_running=_blocking,
+        _local_testing_mode=_local_testing_mode,
+    )[0]
+
+
+@PublicAPI(stability="beta")
```

Comment:
Let's make this a `DeveloperAPI` -- I'm not comfortable maintaining it as a public API with the regular guarantees on stability (yet)